src/entities/join/composable-stable/composableStableJoin.ts

```diff
@@ -0,0 +1,215 @@
+import { encodeFunctionData } from 'viem';
+import { Token } from '../../../entities/token';
+import { TokenAmount } from '../../../entities/tokenAmount';
+import { Address } from '../../../types';
+import { BALANCER_VAULT, MAX_UINT256, ZERO_ADDRESS } from '../../../utils';
+import { vaultAbi } from '../../../abi';
+import {
+    BaseJoin,
+    JoinBuildOutput,
+    JoinInput,
+    JoinKind,
+    ComposableStableJoinQueryResult,
+    ComposableJoinCall,
+} from '../types';
+import { AmountsJoin, PoolState } from '../../types';
+import { doQueryJoin, getAmounts, parseJoinArgs } from '../../utils';
+import { ComposableStableEncoder } from '../../encoders/composableStable';
+
+export class ComposableStableJoin implements BaseJoin {
+    public async query(
+        input: JoinInput,
+        poolState: PoolState,
+    ): Promise<ComposableStableJoinQueryResult> {
+        const bptIndex = poolState.tokens.findIndex(
+            (t) => t.address === poolState.address,
+        );
+        const amounts = this.getAmountsQuery(poolState.tokens, input, bptIndex);
+        const amountsWithoutBpt = {
+            ...amounts,
+            maxAmountsIn: [
+                ...amounts.maxAmountsIn.slice(0, bptIndex),
+                ...amounts.maxAmountsIn.slice(bptIndex + 1),
+            ],
+        };
```

Comment:
How about we make `getAmountsQuery` output both `amounts` and `amountsWithoutBpt`?
I think the main function (i.e. query) would be cleaner and easier to read if we moved this type of manipulation to inner functions.

Reply:
Yeah, I was thinking this can be done in a later refactor after this is merged. I've made an Asana already.